crates/red_knot_test/README.md

```diff
@@ -0,0 +1,430 @@
+TOML fenced code block:
+
+````markdown
+```toml
+[tool.knot]
+warn-on-any = true
+
+```py
+from typing import Any
+
+def f(x: Any):  # error: [use-of-any]
+    pass
+```
+````
+
+It should be possible to include a TOML fenced code block in a single test (as shown), or in a
+grouping section, in which case it applies to all nested tests within that grouping section. Configs
+at multiple level are allowed and merged, with the most-nested (closest to the test) taking
+precedence.
+
+### Configuring search paths and kinds
+
+The red-knot TOML config format hasn't been designed yet, and we may want to implement support in
+the test framework for configuring search paths before it is designed. If so, we can define some
+configuration options for now under the `[tool.knot.tests]` namespace. In the future, perhaps some
+of these can be replaced by real red-knot config options, maybe some of them will be kept long term
+as test-specific options.
+
+We should be able to configure the default workspace root to something other than `/src/` using a
+`workspace-root` config option.
+
+We should be able to add a third-party root using the `third-party-root` config option.
+
+We may want to add additional config options for setting additional search path kinds.
+
+Paths for `workspace-root` and `third-party-root` must be absolute.
+
+Relative embedded-file paths are relative to the workspace root, even if it is explicitly set to a
+non-default value using the `workspace-root` config.
+
+### Specifying custom typeshed
+
+Some tests will need to override the default typeshed with custom files. The `[tool.knot.tests]`
+config option `typeshed-root` should be usable for this:
+
+````markdown
+```toml
+[tool.knot.tests]
+typeshed-root = "/typeshed"
+```
+
+This file is importable as part of our custom typeshed, because it is within `/typeshed`, which we
+configured above as our custom typeshed root:
+
+```py path=/typeshed/stdlib/builtins.pyi
+I_AM_THE_ONLY_BUILTIN = 1
+```
+
+This file is written to `/src/test.py`, because the default workspace root is `/src/ and the default
+file path is `test.py`:
+
+```py
+reveal_type(I_AM_THE_ONLY_BUILTIN)  # revealed: Literal[1]
+
+````
+
+A fenced code block with language `text` can be used to provide a `stdlib/VERSIONS` file in the
+custom typeshed root. If no such file is created explicitly, one should be created implicitly
+including entries enabling all specified `<typeshed-root>/stdlib` files for all supported Python
+versions.
+
+### I/O errors
+
+We could use an `error=` config option in the tag string to make a file cause an I/O error on read.
+
+### Asserting on full diagnostic output
+
+The inline comment diagnostic assertions are useful for making quick, readable assertions about
+diagnostics in a particular location. But sometimes we will want to assert on the full diagnostic
+output of a test. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
```

Comment:
```suggestion
output of a tested Python snippet. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
```

Reply:
I'll probably use "embedded Python file" here for consistency